image-engine/src/main/java/com/safehill/kcrypto/image_engine/internal/ImageFilterWorker.kt

```diff
@@ -0,0 +1,215 @@
+package com.safehill.kcrypto.image_engine.internal
+
+import android.content.ContentValues
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.Color
+import android.net.Uri
+import android.os.Environment
+import android.os.SystemClock
+import android.provider.MediaStore
+import androidx.core.app.NotificationChannelCompat
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import androidx.work.CoroutineWorker
+import androidx.work.ForegroundInfo
+import androidx.work.WorkerParameters
+import com.safehill.kcrypto.image_engine.MLModelsRepository
+import com.safehill.kcrypto.image_engine.R
+import com.safehill.kcrypto.image_engine.jni.JNIProgressTracker
+import com.safehill.kcrypto.image_engine.jni.UpscalingEngine
+import com.safehill.kcrypto.image_engine.model.DownloadModelState
+import com.safehill.kcrypto.image_engine.model.ImageFilterArgs
+import com.safehill.kcrypto.image_engine.model.ImageFilterWorkState
+import com.safehill.kcrypto.image_engine.model.MLModel
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.util.UUID
+
+internal class ImageFilterWorker(
```

Comment:
@DikenMaharjan can you review the coroutine lifecycle here? I'd love to hear your opinion and how integrates with the rest of the coroutines in the codebase